Implement ShortCuts in CanvasItemEditor

neikeq opened the pull request:
I just ended up adding shortcuts for Q, W, E, V and Shift+V. Everything else has mouse events involved and `ShortCut` does not seem to support them right now.

I couldn't add the shortcuts directly to the buttons because it wouldn't check if the key was pressed and if it was not an echo key event... (also it would have been affected by #5174).

neikeq commented on tools/editor/plugins/canvas_item_editor_plugin.cpp:
```diff
@@ -3279,26 +3283,37 @@ CanvasItemEditor::CanvasItemEditor(EditorNode *p_editor) {
+								TTR("Press ($sc) to Change Pivot, ($sc) to Drag Pivot (while moving).")
+													.replace_first("$sc",sc_change_pivot->get_as_text())
+													.replace_first("$sc",sc_drag_pivot->get_as_text())+"\n"+
+								TTR("Alt+RMB: Depth list selection"));
```
Hope this is not too much of a hack...

akien-mga replied:
Wouldn't it be better to use `vformat`?
```
vformat(TTR("Press (%s) to Change Pivot, (%s) to Drag Pivot (while moving)."), sc_change_pivot->get_as_text(), sc_drag_pivot->get_as_text())
```

neikeq replied:
@akien-mga Yes, but it's not valid anyway.

akien-mga replied:
Why that?

neikeq replied:
Read @reduz comment below. We want it to change if the shortcut is changed. There are two possible solutions:
- Extending ShortCut to check if the event is pressed or echo.
- Add a flag to BaseButton to disable the event handling while still keeping the ShortCut.
The second option is the easy one, but the first one would be better IMO.

akien-mga replied:
Ping about using `vformat` proper instead of doing manual replacements on a translated strings.

neikeq replied:
That's the part that still have the problem mentioned in https://github.com/godotengine/godot/pull/5200#issuecomment-233048896. I think the best will be to just leave it as it was before, forcing the <kbd>V</kbd> hotkey. WDYT?

akien-mga replied:
Well I don't have a strong opinion on this, but what I comment about is strictly the syntax. The problem you mentioned should be the same whether replacing `$sc` strings with voodoo magic or using the built-in `vformat` helper function with the `%s` placeholders translators are familiar with, no?

akien-mga replied:
IIUC the tooltip won't get updated when the hotkey is changed (but it would be after an editor restart I guess?), is that right? If so, the current behaviour is still OK-ish and we could open an issue to work on the broader matter instead of holding this PR for another month :)

neikeq replied:
The other option is to listen on settings_changed signal to update the tooltip.

neikeq commented:
Well, [this shortcuts](https://github.com/godotengine/godot/blob/f7d0110cdf61fff70bcb66b50a5274e5d96a6c76/tools/editor/plugins/canvas_item_editor_plugin.cpp#L1835-L1857) could have been added too, but it would make the code a bit ugly :/

reduz commented:
`move_button->set_tooltip(TTR("Move Mode")+" ("+sc_move_mode->get_as_text()+")");`

this has the disadvantage of not changing if the shortcut is changed, but button should add the tooltip by itself shouldn't it?

neikeq commented:
@reduz dammit, I didn't notice that... :/ The button cannot change it because it does not have any shortcut. I explain in the PR description that I could not do that because ShortCut won't check `p_ev.key.pressed && !p_ev.key.echo`, which is required for these buttons.

neikeq commented:
If ShortCut will be leaved as is, a possible alternative fix could be adding a flag to BaseButton to disable the shortcut. The button would still have the shortcut, but it wouldn't handle the events.

akien-mga commented:
CC @djrm who wanted to start working on CanvasItemEditor IIUC.

neikeq commented:
@akien-mga I already talked to him yesterday, he will implement everything except Q, W, E, V and Shift+V, and I will rebase this PR implementing those when this limitation with ShortCut is solved.

djrm commented:
I already have it but travis ci was burning so i wait

akien-mga commented:
Don't wait on travis, it seems broken today.

reduz commented:
is this OK to merge then?

neikeq commented:
@reduz I will have a look at it on Tues/Wed.

reduz commented:
ah, we wanted to release a 2.1 beta tonight, guess this can go for RC or
beta2 then

On Sun, Jul 10, 2016 at 2:31 PM, Ignacio Etcheverry <
notifications@github.com> wrote:

> @reduz <https://github.com/reduz> I will have a look at it on Tues/Wed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/5200#issuecomment-231600658>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AF-Z2waPP2W7LuUq7lmI_GMKDhDqucBxks5qUSx9gaJpZM4I0tb1>
> .
>

neikeq committed:
Implement missing ShortCuts in CanvasItemEditor

neikeq commented:
Updated the commit. Added a `echo` check to BaseButton which fix some of the issues. There is still one more though, which is the displaying of `Change Pivot` and `Drag Pivot` in the tooltip of the `Select Mode` button. Should I just remove the shortcuts for these two options and leave them as they were before?

neikeq commented:
Rebased. I decided not to use ShortCut with _"Change Pivot"_ and _"Drag Pivot"_. This PR is completed.
